### ISSUE TITLE
Running local server on non-localhost causes breakages

### ISSUE THREAD

**Issue by bradfrizzell** [ampproject/amphtml#12054]
Steps to reproduce:
```
gulp --host=<insert name of host machine>
```
navigate to hostname:8000/examples/ads.amp.html
Select doubleclick for ad network.
Notice how all the ads 404

The issue is that non-localhost is not being identified as localDev in mode.js, which is instead hardcoding localhost in its check for localdev mode.

**Comment by erwinmombay** [comment]
looking

**Comment by rsimha-amp** [comment]
I believe this is stemming from the hardcoded check for `localhost` here: https://github.com/ampproject/amphtml/blob/master/src/mode.js#L79

**Comment by rsimha-amp** [comment]
Another option is to force `localDev` via the mechanism here: https://github.com/ampproject/amphtml/blob/master/src/mode.js#L75 (Not sure where in `gulp watch` or `gulp serve` this can be inserted, though.)

**Comment by bradfrizzell** [comment]
this also complicates it: https://github.com/ampproject/amphtml/blob/master/src/3p-frame.js#L235

**Comment by rsimha-amp** [comment]
@erwinmombay Are you working on a fix? If not, I'm happy to pick this up.

**Comment by rsimha-amp** [comment]
@bradfrizzell @erwinmombay I just sent out a fix for this in #12152. Once merged, the following will no longer result in 404s  when you load `hostname.domain.com:8000/examples/ads.amp.html`:

```
gulp clean
export AMP_TESTING_HOST=<hostname.domain.com>
gulp build
gulp serve --host=<hostname.domain.com>
```

Note that for the ads to actually be served in this scenario, you'll probably have to change how `__amp_source_origin` works, but I'll leave it to the a4a team to determine the behavior.

**Comment by rsimha-amp** [comment]
After merging #12152, a couple of hidden bugs in `examples/ads.amp.html` will be exposed:

1. When `localDev` is set to true in the AMP runtime, the page looks like this:

![image](https://user-images.githubusercontent.com/26553114/33095481-d3b7a9b4-ced1-11e7-98b6-c009f850f9c4.png)

2. When `localDev` is set to true, and `thirdPartyFrameHost` and `thirdPartyFrameRegex` are set to a non-localhost machine name, the page looks like this:

![image](https://user-images.githubusercontent.com/26553114/33095503-e2cc4900-ced1-11e7-8032-354c12922857.png)

A workaround for 1 is to run the following commands, following which the page will load correctly:
```
gulp build
gulp prepend-global --target dist/amp.js --{prod|canary}
gulp serve
```

@bradfrizzell assigning this back to you for further investigation.

**Comment by rsimha-amp** [comment]
I have a fix (for item 1 in https://github.com/ampproject/amphtml/issues/12054#issuecomment-346154439) out for review at #12458

**Comment by rsimha-amp** [comment]
@bradfrizzell Assigning this back to you for the case where `localDev` is true, and the ads are being served from a non-localhost server. Let me know what you find out.

**Comment by lannka** [comment]
@rsimha-amp shall we completely remove any hardcoded `localhost` check in AMP (like [here](https://github.com/ampproject/amphtml/blob/master/src/mode.js#L79))? Instead, we can have our local server to prepend "AMP_CONFIG.localDev = true" to the served binary.

That makes sure AMP runtime can be run under any domain in dev mode, which is sometimes very helpful if we want to locally fake the CDN domain and test CORS related logic (like A4A fast fetch).

**Comment by rsimha-amp** [comment]
@lannka I originally removed the `localhost` check you pointed to, but it had to be restored in #11783 by @erwinmombay because it broke some of our test pages.

I do agree though, that it makes sense to simplify the logic in `mode.js`.

**Comment by lannka** [comment]
```js
  const FORCE_LOCALDEV = !!(self.AMP_CONFIG && self.AMP_CONFIG.localDev);
  const AMP_CONFIG_3P_FRAME_HOST = self.AMP_CONFIG &&
      self.AMP_CONFIG.thirdPartyFrameHost;
  const isLocalDev = IS_DEV 
      && !!(win.location.hostname == 'localhost' 
          || (FORCE_LOCALDEV && win.location.hostname == AMP_CONFIG_3P_FRAME_HOST) 
          || (win.location.ancestorOrigins && win.location.ancestorOrigins[0] && startsWith(win.location.ancestorOrigins[0], 'http://localhost:'))
      )
      &&
      // Filter out localhost running against a prod script.
      // Because all allowed scripts are ours, we know that these can only
      // occur during local dev.
      (!win.document || !!win.document.querySelector(developmentScriptQuery));
```

This is the current `isLocalDev` logic. 
@erwinmombay can you talk about the necessity of all those conditions?

I want to simplify this to:

```js
 const isLocalDev = IS_DEV && FORCE_LOCALDEV;
```

And let our gulp build or gulp server to always append `AMP_CONFIG.isLocalDev = true`

**Comment by erwinmombay** [comment]
@lannka i honestly don't remember why we need all the `location` checks (probly a time when we didnt have force local dev yet and even no amp config yet). but if we can make it simpler, go for it.

**Comment by lannka** [comment]
Thanks @erwinmombay .

I tested the proposed change, once we prepend AMP_CONFIG to amp.js, the ads examples start to work.

However, right now I have to do:
```
gulp build --config=canary
gulp serve
```

It does not work if I ran the local server using:
```
gulp
```
@rsimha-amp can we make build with "--config=canary" the default?

**Comment by rsimha-amp** [comment]
We use `--config=prod` by default. Why does that not work?

**Comment by rsimha-amp** [comment]
@lannka do both sets of steps in https://github.com/ampproject/amphtml/pull/12458#issue-281935749 work with your change? What about the steps in https://github.com/ampproject/amphtml/issues/12054#issuecomment-346109474? And what about Heroku? I believe some of those conditions were put in to make all these scenarios work.

I like your fix though, so let me know how I can help to make it real :)

**Comment by lannka** [comment]
> We use --config=prod by default. Why does that not work? What would it take to make that work too?

I think `--config=prod` is not turned on when running `gulp`. I looked at the output at `dist/amp.js`, it does not have the AMP_CONFIG prepend.

I think the fix should be in gulpfile.js. Currently the default `gulp` command is an alias of `gulp watch`, we might have `gulp watch` to support `--config=prod`.

**Comment by rsimha-amp** [comment]
@lannka Good catch. Sent you #12665 to fix the behavior of `gulp` and `gulp watch`.

**Comment by rsimha-amp** [comment]
@lannka made it so that `gulp` and `gulp watch` also enable `localDev` by default in `AMP_CONFIG`. I think we can now implement the simplification you suggested. Sent you #12667 with the fix. PTAL.

**Comment by rsimha-amp** [comment]
@bradfrizzell Now that the `localDev` logic is simplified, there is only one item left to sort out. It turns out that when you run...
```
export AMP_TESTING_HOST=<hostname>
gulp --host=<hostname>
```
... and navigate to `http://<hostname>:8000/examples/ads.amp.html`, the ads don't load because the page is trying to fetch them from `http://ads.localhost:8000/dist.3p/current` and not `http://<hostname>:8000/dist.3p/current`.

I think if we change https://github.com/ampproject/amphtml/blob/master/build-system/app.js#L610 and https://github.com/ampproject/amphtml/blob/master/build-system/app.js#L637 to replace `localhost` in the ads URL with the `<hostname>`, the problem originally reported in this issue should be fixed.

WDYT?

**Comment by rsimha-amp** [comment]
Reopening based on #12666 and https://github.com/ampproject/amphtml/pull/12673#issuecomment-356029846

**Comment by rsimha-amp** [comment]
Paging @erwinmombay @lannka @ampproject/a4a :)

After debugging, I'm seeing that `validateAllowedTypes()` in `3p/integration.js` isn't seeing the `thirdParty*` overrides in `self.AMP_CONFIG` that get applied when we set `AMP_TESTING_HOST`. However, `getAdsLocalhost()` in `src/3p-frame.js` is correctly seeing them.

Presumably, this is because the AMP config overrides aren't being applied to the 3rd party frame, which is probably why non-localhost scenarios never did work.

Do any of you know of a recommended way to apply testing overrides to `AMP_CONFIG` such that third party frames use those values too?

**Comment by rsimha-amp** [comment]
Turns out we need to write `AMP_CONFIG` to the 3p frame as well, for local development with a non-localhost server. Fix available in #12703.
